### ISSUE TITLE
"Cache directory creation problem" and "source image could not be found" with v1.7

### ISSUE THREAD

**Issue by davidromani** [liip/LiipImagineBundle#846]
Hello.

After upgrade this bundle from 1.6 to 1.7 my project doesn't works well. 

In my local environment everything works well. My "web/media/cache" dir exists with all permissions. But when I deploy my app to a production server with Capifony, it seems that this bundle can't create "cache" directory.

Is there some change related with directory creation in this update? In my deployment server I only have "web/media" dir with all permissions, the "web/media/cache" is not deployed...

**Comment by antoligy** [comment]
I don't believe that there's been any changes to the creation (or lack thereof) of the `web/media/cache` directory.  I'll take a look to double-check...

(For reference - https://github.com/liip/LiipImagineBundle/compare/1.6.0...1.7.0)

**Comment by adaniloff** [comment]
Hello there,

+1 with davidromani.

We got a problem with an update in the `Binary/Loader/FileSystemLoader.php` file.
I got the "Source image invalid "%s" as it is outside of the defined root path" error message since 1.7 upgrade.

Tried to downgrade to 1.6.* and it worked fine again.

It seems to be an environment problem (maybe because of symlinks), we're actually trying to figure it out; because it works perfectly on our local env but does not on production.

**Comment by raphaChoquet** [comment]
Hello,

The error is due to symbolic links created by 'assets:install --symlink'.
I made a pull request to fix it :  
https://github.com/liip/LiipImagineBundle/pull/848

**Comment by picks44** [comment]
Thanks @raphaChoquet! Hope there'll be a new release with this fix soon, I had to downgrade to 1.6.

**Comment by antoligy** [comment]
Unfortunately this PR has broken the testing suite and so I'm not comfortable with merging and releasing it just yet.  However I'll prioritise it, so as soon as we can be confident it won't break then we will create a patch level release.

**Comment by raphaChoquet** [comment]
Yes, tests has broken because program doesn't replace paths containing '../' in $path.
I think we need to replace realpath function by other logic to keep symbolic link.

**Comment by robfrawley** [comment]
@raphaChoquet We've cover this in multiple issues prior: you need to set a proper `data_root` value in your bundle configuration to allow for the **real path** once the symbolic link is resolved.

**Comment by antoligy** [comment]
Thanks for the clarification @robfrawley, I remember that being a particular pain-point in earlier issues.  I wasn't actually aware of the need to configure data_root either, so I wonder if it's worth us documenting that to avoid more issues of this nature?

**Comment by robfrawley** [comment]
@antoligy It is documented here: https://github.com/liip/LiipImagineBundle#images-outside-data-root

We could possibly highlight it more and add some clarity. If I remember correctly, that existed almost exactly as it currently is before I submitted the documentation re-write; we could expand upon that and provide some additional clarity, specifically point to the common error message(s) people get when they do not have a proper `data_root` value set.

**Comment by robfrawley** [comment]
Also, we don't mention the difficulties people encounter when they utilize symbolic links, so perhaps that should be explicitly spelled out in the documentation as well. I won't have any time to work on it before the weekend, but if no one else gets around to it before, I'll happily expand on that sometime this coming weekend.

**Comment by robfrawley** [comment]
@davidromani @adaniloff @raphaChoquet @picks44 The issue you are encountering is the following: prior releases didn't consistently chroot the file within the resolved root path, which was the intended behavior, it was simply written in an easily (and regularly) bypassed manned. Moving forward, the code requires images reside in the root configured. This means that if an image lives in `/path/a` and you have the following symbolic link `/path/symlink` pointing to `/path/a`, the `data_root` bundle configuration value must be set to `/path/a`. Does this resolve your issues?

**Comment by adaniloff** [comment]
@robfrawley Reading the doc, it seems that we can only configure one data_root. It could be a pain in the *ss. For example, I got src/Foo1Bundle and src/Foo2Bundle having their own resources, how should I do?

Configure all the SRC/ as data_root path? Because it seems bad to do so to me.

Also, since it's an update from 1.6 to 1.7 only, I think the presence of this parameter should be checked and forced. Otherwise, a lot of people will one day composer update their production, and have this trouble without understanding why.

I may be wrong so please explain me the how and why if so.

Cheers

**Comment by robfrawley** [comment]
@adaniloff Reference my [latest comment from the PR](https://github.com/liip/LiipImagineBundle/pull/848#issuecomment-272523363). Would allowing an array of data roots solve your problem while ensuring the security of the system?

**Comment by robfrawley** [comment]
@adaniloff Can you patch your imagine-bundle with PR #851 and configure the multiple data roots needed by your setup and see if this resolves your issue? To do so, ensure you've upgraded your imagine-bundle to `v1.7` and enter the bundle root `cd vendor/liip/imagine-bundle` and call the patch command `patch -p1 < curl https://patch-diff.githubusercontent.com/raw/liip/LiipImagineBundle/pull/851.diff` after which you can define multiple data roots using:
```yml
liip_imagine:
  loaders:
    default:
      filesystem:
        data_root: 
          - "/your/data/root/a"
          - "/your/data/root/b"
```
Does this resolve your issue?

**Comment by adaniloff** [comment]
@robfrawley Thanks for your replies and explanations. I can't try it right now as I won't have my computer with me until Monday. 

Gonna keep you informed about https://github.com/liip/LiipImagineBundle/pull/851.

I'd like to know if it is also possible to add a user notice (or warning) when upgrading from 1.6.x to 1.7+ as following :
- the user do a `composer update` that change the version of this bundle
- in the DependencyInjection/Factory/Loader/FileSystemLoaderFactory.php, you check the presence of an optionnal parameter (DependencyInjection/Configuration.php should also be change according to accept a this parameter as an optionnal boolean) like that : 
```
class FileSystemLoaderFactory implements LoaderFactoryInterface
{
    /**
     * {@inheritdoc}
     */
    public function create(ContainerBuilder $container, $loaderName, array $config)
    {
        $loaderDefinition = new DefinitionDecorator('liip_imagine.binary.loader.prototype.filesystem');
        if (false == isset($config['data_root']) || empty($config['data_root'])
            && (false == isset($config['disable_notice']) || true !== (bool)$config['disable_notice'])) {
            trigger_error('You must configure at least one data_root path or disable notice from 1.7', E_USER_NOTICE);
        }
        $loaderDefinition->replaceArgument(2, $config['data_root']);
        $loaderDefinition->addTag('liip_imagine.binary.loader', array(
            'loader' => $loaderName,
        ));
        $loaderId = 'liip_imagine.binary.loader.'.$loaderName;
        $container->setDefinition($loaderId, $loaderDefinition);
        return $loaderId;
    }

```

I made this up quickly because I won't be able to do it properly before, at least Monday as I said. This example is just to show you what I'd want, if possible :).

The idea, is to explicitly show the developer the need to check the doc about the data_root arguments, without impacting the production environments. Also, the optionnal parameter is here to let a developer bypass the configuration, without triggering error.

My concern is, as the images are cached, when a dev. will `composer update`, he won't always see the problem occurs because Liip won't resolve already cached images. So he could update the production env and then spend to much time trying to understanding what's up.

I hope I made myself clear, it's a bit late over there and english is not my native language.

Regards.

**Comment by robfrawley** [comment]
@adaniloff I absolutely understand what you are trying to convey in terms of providing notice to the user, but I'm not sure your solution is indicative of the broader usage of this bundle: *most people don't set `data_root` and never need to do so.* As such, IMHO, it seems counter-intuitive to require they set a value for this option (when they normally would not) *or* set a configuration option whose sole purpose is to silence a notice (the `disable_notice` key in your example).

Perhaps adding a composer hook for this upgrade cycle that provides an explicit notice to the user? Any composer buffs out that that may have an alternate solution?

**Comment by adaniloff** [comment]
@robfrawley 
I was convinced that this would be the case of most of the people (I mean to use symlinks for images etc), but reading your previous message it seems not to be.

A composer hook could do the trick I guess. Outputing a big warning / danger message :)

Regards

**Comment by robfrawley** [comment]
@adaniloff Here is a notice that is pulled from the UPGRADE.md file (it matches the version hesder to the version being installed/upgraded). Thoughts?
![liip-imagine-composer-notice](https://cloud.githubusercontent.com/assets/3967713/21959150/ae1a1b4a-da8c-11e6-9453-195f53c30b07.png)

**Comment by antoligy** [comment]
I really like this approach and honestly now wish that other bundles were doing it too!  (It'd save me a *lot* of time 😂)

**Comment by adaniloff** [comment]
@robfrawley I already love it. Thanks :)

**Comment by robfrawley** [comment]
Looks like the response is positive so far, so I'll work on cleaning this up and getting it ready for submission. Note that it is "interactive-aware" and doesn't prompt the user to confirm the notice when Composer is run in non-interactive mode (for automated scripts or other non-interactive environments). I'll get a PR submitted later tonight.

**Comment by robfrawley** [comment]
See PR #853 for composer update notice hook.

**Comment by adaniloff** [comment]
@robfrawley Hi!
So you decided to abort the composer update hook part right ?
Could you be more explicit on the "why" ? I'm not quite sure that I have understood it.
Also, do you think of any other way to do something similar?

Sorry for bothering you that much :)

Regards

**Comment by robfrawley** [comment]
@adaniloff The composer hook was abandoned because the `scripts` key of a composer configuration is ignored unless it is part of the **root** `composer.json` file. See their [documentation on the subject](https://getcomposer.org/doc/articles/scripts.md#what-is-a-script-) for additional confirmation. What this means is that it would only be executed for developers of the bundle who are running composer explicitly for the package, and it would not do anything for people who **require** the package in their own projects.

As for an alternative, I have yet to think of a solid way to handle this, but it is still on my mind, and I am open to ideas. That said, I am still against pursuing your original idea (re: https://github.com/liip/LiipImagineBundle/issues/846#issuecomment-272584682).

**Comment by robfrawley** [comment]
Oops: didn't mean to close this (must have hit "comment and close" instead of just "comment") -- my bad!

**Comment by robfrawley** [comment]
Version [1.7.1](https://github.com/liip/LiipImagineBundle/releases/tag/1.7.1) released with multiple data roots support.

**Comment by robfrawley** [comment]
@davidromani Is your issue resolved?

**Comment by davidromani** [comment]
@robfrawley sorry, but I don't undestand why allowing multiple data roots can fix my problem... I have a legacy project deployed with Capifony and every release creates a new directory. 

Check chapter "4.Setup server" here http://capifony.org to see my directory structure. I usually share user uploads directory between deployments, and my web/media/cache dir is **not** shared.

So I only have a unique data root config like this:

```
liip_imagine:
    loaders:
        default:
            filesystem:
                data_root: 
                    - "%kernel.root_dir%/../web/"
```

With this solution I must add a new data_root dir for each deploy? It's quite cumbersome...

**Comment by antoligy** [comment]
You should be able to specify a data root pointing to the current
directory, does that work?

**Comment by davidromani** [comment]
@antoligy I've tried to config this setting and it **doesn't** work.

```
liip_imagine:
    loaders:
        default:
            filesystem:
                data_root:
                    - "/path/to/my-web-project/current/web/"
```

And my directory structure is:

```
$ pwd
/path/to/my-web-project
$ ls -lh
total 8.0K
lrwxrwxrwx 1 flux flux   54 Jan 30 09:39 current -> /path/to/my-web-project/releases/20170130083913
drwxrwxr-x 5 flux flux 4.0K Jan 30 09:39 releases
drwxrwxr-x 5 flux flux 4.0K Nov 26  2015 shared
```

**Comment by robfrawley** [comment]
@davidromani That won't work. You must add the symlink *resolved* paths to your `data_root` (in your case: `/path/to/my-web-project/releases/20170130083913`. Alternatively, you can use bind mounts or other means.

**Comment by davidromani** [comment]
@robfrawley so it's not a solution for me because every release the path is changing...

I know that is not a related problem with this bundle, so you can close this issue if you want. Thanks.

**Comment by robfrawley** [comment]
Again, check out bind mounts, or hard links.

**Comment by antoligy** [comment]
Actually - I am not sure that those are so easy to set up when working with Capistrano, especially where some webservers (at least, Apache) relies on the FollowSymlinks directive in order to do rolling deploys.
Wondering if it's worth some kind of Composer script that can automatically propagate data root configurations at build/deploy time?  (also not ideal as this kills the immutability of a deployment, but I guess there's a point at which we have to draw a line)

**Comment by robfrawley** [comment]
Take, for example, hard links. You could set up the following directory structure:
```
current/ -> ./releases/20170130083913 (normal synlink, handled by Capistrano)
releases/
application/ -> ./current (hardlink, created yourself)
```

And use `application/` as the data root for Apache, Imagine Bundle, etc.

```yml
liip_imagine:
    loaders:
        default:
            filesystem:
                data_root:
                    - "/path/to/web/application/web/"
```

Given the above, because `application/` is a hard link, it won't be resolved beyond itself, as to the filesystem it appears as a normal directory, despite the fact that it points to a symbolic link that points to the real folder.

**Comment by robfrawley** [comment]
@antoligy Do we want to offer a toggle to disable security checks? And simply allow whatever paths are passed? (Disabled by default, of course).

**Comment by robfrawley** [comment]
@davidromani Lastly, you could very easily write your own loader that does or does not perform whatever checks you want for your specific situation. I mean, the interface isn't particular difficult to implement, so it'd take only a few minutes:

```php
interface LoaderInterface
{
    /**
     * Retrieve the Image represented by the given path.
     *
     * The path may be a file path on a filesystem, or any unique identifier among the storage engine implemented by this Loader.
     *
     * @param mixed $path
     *
     * @return \Liip\ImagineBundle\Binary\BinaryInterface|string An image binary content
     */
    public function find($path);
}
```

**Comment by robfrawley** [comment]
@davidromani What is the exact, complete, exception you get when using the `current/` directory as the data root?

**Comment by robfrawley** [comment]
Resolved via [1.7.2 release](https://github.com/liip/LiipImagineBundle/releases/tag/1.7.2)!

**Comment by akrz** [comment]
For anyone looking the less secure option:

liip_imagine :
  loaders:
    default:
      filesystem:
        locator: **filesystem_insecure**
